neural_modelling/src/bit_field_expander/bit_field_expander.c

```diff
@@ -0,0 +1,455 @@
+//! \return bool that states if it was successful at generating the bitfield
+bool generate_bit_field(){
+
+    // write how many entries (thus bitfields) are to be generated into sdram
+    log_debug("update by pop length");
+    bit_field_base_address->n_filters = population_table_length();
+
+    // location where to dump the bitfields into (right after the filter structs
+    address_t bit_field_words_location =
+        (address_t) &bit_field_base_address->filters[
+            bit_field_base_address->n_filters];
+    log_info("words location is %x", bit_field_words_location);
+    int position = 0;
+
+    // iterate through the master pop entries
+    log_debug("starting master pop entry bit field generation");
+    for (uint32_t master_pop_entry=0;
+            master_pop_entry < population_table_length();
+            master_pop_entry++){
+
+        // determine keys masks and n_neurons
+        spike_t key = population_table_get_spike_for_index(master_pop_entry);
+        uint32_t mask = population_table_get_mask_for_entry(master_pop_entry);
+        uint32_t n_neurons = _n_neurons_from_key(key);
+
+        // generate the bitfield for this master pop entry
+        uint32_t n_words = get_bit_field_size(n_neurons);
+
+        log_debug("pop entry %d, key = %d, mask = %0x, n_neurons = %d",
+                  master_pop_entry, (uint32_t) key, mask, n_neurons);
+        bit_field_t bit_field = bit_field_alloc(n_neurons);
+        if (bit_field == NULL){
+            log_error("could not allocate dtcm for bit field");
+            return false;
+        }
+
+        // set the bitfield to 0. so assuming a miss on everything
+        clear_bit_field(bit_field, n_words);
+        log_debug("cleared bit field");
+
+        // update sdram with size of this bitfield
+        bit_field_base_address->filters[master_pop_entry].key = key;
+        log_debug(
+            "putting master pop key %d in entry %d",
+            key, master_pop_entry);
+        bit_field_base_address->filters[master_pop_entry].n_words = n_words;
+        log_debug("putting n words %d in entry %d", n_words, master_pop_entry);
+
+        // iterate through neurons and ask for rows from master pop table
+        log_debug("searching neuron ids");
+        for (uint32_t neuron_id=0; neuron_id < n_neurons; neuron_id++){
+
+            // update key with neuron id
+            spike_t new_key = (spike_t) (key + neuron_id);
+            log_debug("new key for neurons %d is %0x", neuron_id, new_key);
+
+            // holder for the bytes to transfer if we need to read sdram.
+            size_t n_bytes_to_transfer;
+
+            // used to store the row from the master pop / synaptic matrix,
+            // not going to be used in reality.
+            address_t row_address;
+            bool bit_found = false;
+            if (population_table_get_first_address(
+                    new_key, &row_address, &n_bytes_to_transfer)){
+                log_info("%d", neuron_id);
```

Comment:
Not a review - yet - but this (and just below on line 385) should be a log_debug(...).  This causes e.g. the mouse brain model to RTE due to the IOBUF filling up (I think).